


rule sort_bam:
    input:
        alignment=expand("{out}/alignment/{sample}_trimmed_filt_merged.1_bismark_bt2_pe.bam",out=config["output_dir"],sample=SAMPLES),
    output:
        alignmentSorted=expand("{out}/alignment/{{sample}}_sorted.bam",out=config["tmpdir"])
    conda:
        "../env/samtools.yaml"
    shell:
        "samtools sort {input.alignment} > {output.aligmentSorted}"

        
# call SNPs from recalibrated alignments
rule call_SNPs_denovo:
    input:
        alignment=expand("{out}/alignment/{sample}_sorted.bam",out=config["tmpdir"],sample=SAMPLES)
        reference=expand("{path}/output_denovo/NNNNref/ref.fa",path=config["output_dir"])
        referenceDict=expand("{path}/output_denovo/NNNNref/ref.dict",path=config["output_dir"])
    output:
        perIndVCF=expand("{path}//snp_calling/{{sample}}_snp.raw.vcf.out",path=config["output_dir"]),
        bisSNP_meth=expand("{path}/output/methylation_calling/bis-SNP/{{sample}}_cpg.raw.vcf")
    params:
        sample="{sample}"
    log:
        "/log/{sample}_call.log"
    conda:
        "../env/env_bis-snp0.82.yaml"
    threads: 1
    shell:
        "(bis-snp -T BisulfiteGenotyper -R {input.reference} -I {input.alignment} -vfn1  -vfn2 {output.perIndVCF} -stand_call_conf 20 -stand_emit_conf 0 -toCoverage 1000 -mmq 20 -mbq 20) 2> {log}"

